aggregator/election_mgr.go

```diff
@@ -0,0 +1,518 @@
+	mgr.metrics.resignNotFollower.Inc(1)
+	mgr.logError("resign failed", err)
+	return err
+}
+
+func (mgr *electionManager) Close() error {
+	mgr.Lock()
+	defer mgr.Unlock()
+
+	if mgr.state != electionManagerOpen {
+		return errElectionManagerExpectOpen
+	}
+	close(mgr.doneCh)
+	mgr.state = electionManagerClosed
+	return nil
+}
+
+func (mgr *electionManager) startElectionLoop() {
+	var campaignStatusCh <-chan campaign.Status
+	notDone := func(int) bool {
+		select {
+		case <-mgr.doneCh:
+			return false
+		default:
+			return true
+		}
+	}
+
+	for {
+		if campaignStatusCh == nil {
+			mgr.RLock()
+			state := mgr.state
+			resignWg := mgr.resignWg
+			mgr.RUnlock()
+
+			// NB(xichen): need to wait for the result of the leader service resign call
+			// because resigning is done in a different goroutine and when/whether we need
+			// to restart the campaign depends on the result of the resign call.
+			if state == electionManagerResigning {
+				resignWg.Wait()
+			}
+
+			// NB(xichen): campaign retrier retries forever until either the Campaign call succeeds,
+			// or the election manager has resigned.
+			if err := mgr.campaignRetrier.AttemptWhile(notDone, func() error {
+				var err error
+				campaignStatusCh, err = mgr.leaderService.Campaign(mgr.electionKey, mgr.campaignOpts)
+				if err == nil {
+					return nil
+				}
+				mgr.metrics.campaignCreateErrors.Inc(1)
+				mgr.logError("error creating campaign", err)
+				return err
+			}); err == xretry.ErrWhileConditionFalse {
+				return
+			}
+		}
+
+		var (
+			campaignStatus campaign.Status
+			ok             bool
+		)
+		select {
+		case campaignStatus, ok = <-campaignStatusCh:
+			// If the campaign status channel is closed (e.g., because session has expired, or
+			// we have resigned from the campaign, or there are issues with the underlying etcd
+			// cluster), we restart the campaign loop.
+			if !ok {
+				campaignStatusCh = nil
+				continue
+			}
+		case <-mgr.doneCh:
+			return
+		}
+		mgr.processStatusUpdate(campaignStatus)
+	}
+}
+
+func (mgr *electionManager) processStatusUpdate(campaignStatus campaign.Status) {
+	if campaignStatus.State == campaign.Error {
+		mgr.metrics.campaignErrors.Inc(1)
+		mgr.logError("error campaigning", campaignStatus.Err)
+		return
+	}
+
+	currState := mgr.ElectionState()
+	newState := FollowerState
+	if campaignStatus.State == campaign.Leader {
+		newState = LeaderState
+	}
+	mgr.changeElectionState(currState, newState)
+}
+
+func (mgr *electionManager) changeElectionState(currState, newState ElectionState) {
+	if currState == newState {
+		if newState != LeaderState {
+			return
+		}
+		// We are in the middle of verifying leader and changing from leader to follower
+		// and we received an update that indicates the current instance is now the leader,
+		// therefore we cancel the change from leader to follower and instead set the
+		// election state to leader.
+		if mgr.cancelInProgressChange() {
```

Comment:
Again I don't see any in progress change that happens during a resign?  i.e. cancel fn never gets set.